tests/apps/test_cost.py

```diff
@@ -91,7 +94,288 @@ def test_grad_threshold(self, k):
+    """Fully connected and non-normalized adjacency matrix"""
+    return np.ones((dim, dim))
+
+
+@pytest.fixture
+def vgbs(A, n_mean, embedding, threshold):
+    """VGBS for the fully connected adjacency matrix using the exponential embedding"""
+    return train.VGBS(A, n_mean, embedding, threshold)
+
+
+@pytest.fixture
+def params(dim):
+    """Trainable parameters initialized as a slight perturbation from all zeros"""
+    return np.array([0.01 * i for i in range(dim - 1)])
+
+
+@pytest.mark.parametrize("dim", [4])
+@pytest.mark.parametrize("n_mean", [1])
+@pytest.mark.usefixtures("vgbs")
+class TestStochastic:
+    """Tests for the class ``train.Stochastic``"""
+
+    @pytest.mark.parametrize("threshold", [False])
+    def test_h_reparametrized(self, vgbs, dim, params):
+        """Test that h_reparametrized behaves as expected by calculating the cost function over
+        a fixed set of PNR samples using both the reparametrized and non-reparametrized methods"""
+        cost_fn = train.Stochastic(h, vgbs)
+        possible_samples = list(itertools.product([0, 1, 2], repeat=dim))
+
+        probs = [vgbs.prob_sample(params, s) for s in possible_samples]
+        probs_init = [vgbs.prob_sample(np.zeros(dim - 1), s) for s in possible_samples]
+
+        cost = sum([probs[i] * h(s) for i, s in enumerate(possible_samples)])
+        cost_reparam = sum(
+            [
+                probs_init[i] * cost_fn.h_reparametrized(s, params)
+                for i, s in enumerate(possible_samples)
+            ]
+        )
+
+        assert np.allclose(cost, cost_reparam)
+
+    @pytest.mark.parametrize("threshold", [False])
+    def test_h_reparametized_example(self, dim, vgbs, params, embedding):
+        """Test that h_reparametrized returns the correct value when compared to working the
+        result out by hand"""
+        cost_fn = train.Stochastic(h, vgbs)
+        sample = np.ones(dim)
+        h_reparam = cost_fn.h_reparametrized(sample, params)
+        h_reparam_expected = -1.088925964188385
+        assert np.allclose(h_reparam, h_reparam_expected)
+
+    @pytest.mark.parametrize("threshold", [False])
+    def test_evaluate(self, vgbs, dim, params):
+        """Test that evaluate returns the expected value when the VGBS class is preloaded with a
+        dataset where half of the datapoints are zeros and half of the datapoints are ones. The
+        expected result of the evaluate method is then simply the average of h_reparametrized
+        applied to a ones vector and a zeros vector."""
+        n_samples = 10
+        zeros = np.zeros((n_samples, dim))
+        ones = np.ones((n_samples, dim))
+        samples = np.vstack([zeros, ones])
+        vgbs.add_A_init_samples(samples)
+
+        cost_fn = train.Stochastic(h, vgbs)
+        h0 = cost_fn.h_reparametrized(zeros[0], params)
+        h1 = cost_fn.h_reparametrized(ones[0], params)
+        eval_expected = (h0 + h1) * 0.5
+
+        eval = cost_fn(params, 2 * n_samples)
```

Comment:
could leave a small comment here referring to the fact that calling the instance of `Stochastic` calls on `Stochastic.evaluate`